### PR TITLE
Modify Retry Approach to Propagate Errors

### PR DESCRIPTION
**Pull request by dzou**
This modifies the retry logic of the driver to propagate `R2dbcTransientExceptions` if a call fails instead of attempting to retry on the driver-side.

### CODE REVIEW

File: src/main/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtil.java [hunk 1]
```diff
@@ -88,6 +79,10 @@ public void onNext(ResponseT value) {
 
     @Override
     public void onError(Throwable t) {
+      if (isRetryable(t)) {
+        t = new R2dbcTransientResourceException(t);
+      }
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtil.java]: else wrap it into `R2dbcNonTransientException`?

File: src/main/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtil.java [hunk 2]
```diff
@@ -135,6 +130,11 @@ public void onNext(ResponseT response) {
     @Override
     public void onError(Throwable throwable) {
       this.terminalEventReceived = true;
+
+      if (isRetryable(throwable)) {
+        throwable = new R2dbcTransientResourceException(throwable);
+      }
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtil.java]: Same here wrap into `R2dbcNonTransientException`?
Also, consider adding an exception reason.

File: src/main/java/com/google/cloud/spanner/r2dbc/client/GrpcClient.java [hunk 3]
```diff
@@ -171,60 +167,9 @@ public GrpcClient(GoogleCredentials credentials) {
...
-                  .request((int) Math.min(demand, Integer.MAX_VALUE)));
-          }));
-  }
-
-  @VisibleForTesting
-  SpannerStub getSpanner() {
-    return this.spanner;
-  }
-
-  private static final class SinkResponseObserver<ReqT, RespT> implements
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/client/GrpcClient.java]: Nice catch! How did we miss this refactor? Probably tricky merges.

Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/client/GrpcClient.java]: I originally [meant to remove](https://github.com/GoogleCloudPlatform/cloud-spanner-r2dbc/blob/246e6f656ae748a8061af13f3e62b34378dd4009/src/main/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtil.java#L59) the implementation in ObservableReactiveUtil, but forgot.
They are functionally equivalent, I think, so it does not matter which one to keep.

File: src/main/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtil.java (hunk 1 above)
Comment (dzou) [reply src/main/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtil.java]: Done.

File: src/main/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtil.java (hunk 2 above)
Comment (dzou) [reply src/main/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtil.java]: Done.

File: src/main/java/com/google/cloud/spanner/r2dbc/client/GrpcClient.java (hunk 3 above)
Comment (dzou) [reply src/main/java/com/google/cloud/spanner/r2dbc/client/GrpcClient.java]: Done.

**Comment by dzou** [comment]
> If you are keeping the ObservableReactiveUtil implementation, can you remove my comment about "this will go away"?
> Otherwise LGTM; I agree with Mike's suggestion to wrap in R2dbcNonTransientException for the other case.

Gotcha. Donerino.

File: src/test/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtilTest.java [hunk 4]
```diff
@@ -45,57 +47,51 @@ public void unaryCallForwardsError() {
       observer.onError(new IllegalArgumentException("oh no"));
     });
     assertThatThrownBy(() -> mono.block())
-        .isInstanceOf(IllegalArgumentException.class)
```
Comment (elefeint) [inline src/test/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtilTest.java]: Could you keep the check for root exception as `hasCauseInstanceOf()`

Comment (dzou) [reply src/test/java/com/google/cloud/spanner/r2dbc/util/ObservableReactiveUtilTest.java]: Done.

**Pull request merged**